exporters/memory/include/opentelemetry/exporters/memory/in_memory_metric_exporter.h

```diff
@@ -0,0 +1,56 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <atomic>
+#include <vector>
+
+#include "opentelemetry/sdk/metrics/export/metric_producer.h"
+#include "opentelemetry/sdk/metrics/instruments.h"
+#include "opentelemetry/sdk/metrics/push_metric_exporter.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace memory
+{
+
+/// A Push Metric Exporter which accumulates metrics data in memory and allows it to be inspected.
+/// It is not thread-safe.
+class InMemoryMetricExporter final : public sdk::metrics::PushMetricExporter
+{
+  using AggregationTemporalityMap =
+      std::map<sdk::metrics::InstrumentType, sdk::metrics::AggregationTemporality>;
+
+public:
+  /// @param temporality Output temporality as a function of instrument kind.
+  InMemoryMetricExporter(const sdk::metrics::AggregationTemporalitySelector &temporality)
+      : temporality_(temporality)
+  {}
+
+  InMemoryMetricExporter(const InMemoryMetricExporter &)  = delete;
+  InMemoryMetricExporter(const InMemoryMetricExporter &&) = delete;
+  void operator=(const InMemoryMetricExporter &)          = delete;
+  void operator=(const InMemoryMetricExporter &&)         = delete;
+  ~InMemoryMetricExporter() override                      = default;
+
+  sdk::common::ExportResult Export(const sdk::metrics::ResourceMetrics &data) noexcept override;
+  sdk::metrics::AggregationTemporality GetAggregationTemporality(
+      sdk::metrics::InstrumentType instrument_type) const noexcept override;
+  bool ForceFlush(
+      std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept override;
+  bool Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+  const std::vector<sdk::metrics::ResourceMetrics> &GetData() const;
+
+private:
+  std::vector<sdk::metrics::ResourceMetrics> data_{};
```

Comment:
Here the data is owned by the exporter.

Is it possible to follow the same pattern as with spans ?

* have a separate class for the data alone,
* keep a shared pointer to the data from the exporter

See `InMemorySpanExporter` and `InMemorySpanData`.

This makes it easier to inspect the exported data in the application, even after the exporter is shutdown and destroyed for example.

Reply:
Done.